js/src/index.ts

```diff
@@ -6,36 +6,51 @@
 //
 //
 
-import { 
-    Connection, 
+import {
+    Connection,
     Keypair,
-    PublicKey, 
+    PublicKey
 } from '@solana/web3.js';
-import { 
-    NautilusQuery, 
+import { NautilusIdl, NautilusProgram } from './types';
+import {
+    NautilusQuery,
     NautilusTable,
 } from './sql';
-import { NautilusUtils } from './util';
 
-export class Nautilus {
+import { decapitalizeFirstLetter } from './util';
+
+export class Nautilus<Program extends NautilusProgram = NautilusProgram> {
 
     connection: Connection;
-    programs: [PublicKey, string][];
-    defaultProgram: PublicKey | undefined;
-    payer: Keypair | undefined;
+    programId: PublicKey;
+    programs?: {[programName: string]: PublicKey};
+    payer?: Keypair | undefined;
 
-    util: NautilusUtils = new NautilusUtils();
+    readonly idl: NautilusIdl;

```

Comment:
This is cool, but I'm curious if we need to store the whole IDL in the instance, subsequently carrying it around anyone's client bundle?

Reply:
Probably not, I think it could work to only have it as a constructor argument. I'll work on that